### PR TITLE
Annotation Facility

### PR DESCRIPTION
**Pull request by barisegesevgili**
## What this PR does / why we need it:

- Following the standards that are stated at w3c: https://www.w3.org/TR/annotation-model/ annotation facility is implemented. 

## Structure of annotations

- For text annotation to a text content:
```
{
  "@context": "http://www.w3.org/ns/anno.jsonld",
  "id": <Id of the annotation>,
  "type": "Annotation",
  "creator": "https://traderx.company/user/<username>/profile",
  "created": "2015-01-28T12:00:00Z",
  "modified": "2015-01-29T09:00:00Z",
  "body":     
    {
      "type": "Text",
      "value": "Lorem ipsum dolor sit amet"
    }
  "target": {
    "type": "Text",
    "id": <Id of the article>,
    "selector": {
      "type": "TextPositionSelector", (detailed: https://www.w3.org/TR/annotation-model/#text-position-selector)
      "start": 412,
      "end": 795
    }
  }
}
```

- For text annotation to an image content:
```
{
  "@context": "http://www.w3.org/ns/anno.jsonld",
  "id": <Id of the annotation>,
  "type": "Annotation",
  "creator": "https://traderx.company/user/<username>/profile",
  "created": "2015-01-28T12:00:00Z",
  "modified": "2015-01-29T09:00:00Z",
  "body":     
    {
      "type": "Text",
      "value": "Lorem ipsum dolor sit amet"
    }
  "target": {
      "type": "Image",   
      "id": "<id of the article>#xywh=100,100,300,300"
  }
}
```

- For image annotation to a text content:
```
{
  "@context": "http://www.w3.org/ns/anno.jsonld",
  "id": <Id of the annotation>,
  "type": "Annotation",
  "creator": "https://traderx.company/user/<username>/profile",
  "created": "2015-01-28T12:00:00Z",
  "modified": "2015-01-29T09:00:00Z",
  "body":     
    {
      "type": "Image",
      "value": url_of_the_image"
    }
  "target": {
    "type": "Text",
    "id": <Id of the article>,
    "selector": {
      "type": "TextPositionSelector", (detailed: https://www.w3.org/TR/annotation-model/#text-position-selector)
      "start": 412,
      "end": 795
    }
  }
}
```

- For image annotation to a image content:
```
{
  "@context": "http://www.w3.org/ns/anno.jsonld",
  "id": <Id of the annotation>,
  "type": "Annotation",
  "creator": "https://traderx.company/user/<username>/profile",
  "created": "2015-01-28T12:00:00Z",
  "modified": "2015-01-29T09:00:00Z",
  "body":     
    {
      "type": "Image",
      "value": url_of_the_image"
    }
  "target": {
      "type": "Image",   
      "id": "<id of the article>#xywh=100,100,300,300"
  }
}
```

## Endpoint parameters:

- Create an Annotation:
    - Mandatory Fields:
    private String bodyType;  --> either "Text" or "Image" //case sensitive
    private String targetType; --> either "Text" or "Image" //case sensitive
    private Integer articleId; --> id of the targetted article
    private String content;  --> content; could also be a url of an image(we         use the url of the image, when we want image as an annotation) 
    - Fields that are needed base on Target type:
       - If Target type is "Text":
        private Integer posStart; --> starting point of the selected text, if target type is "Image", set as -1
        private Integer posEnd; --> ending point of the selected text, if  target type is "Image", set as -1
        
       - If Target type is "Images":
        private Integer imgX; --> you can use it as you wish, this is the compatible structure with w3c. if target type is "Image", -1 by default
        private Integer imgY; --> you can use it as you wish, this is the compatible structure with w3c. if target type is "Image", -1 by default
        private Integer imgW; --> you can use it as you wish, this is the compatible structure with w3c. if target type is "Image", -1  by default
        private Integer imgH; --> you can use it as you wish, this is the compatible structure with w3c. if target type is "Image", -1 by default

- Update an Annotation:

    - Mandatory Fields:
    private Integer annotationId; --> id of the annotation that will be updated
    private String content;  --> content; could also be a url of an image(we  use the url of the image, when we want image as an annotation) 
    - Fields that are needed base on Target type:
       - If Target type is "Text":
        private Integer posStart; --> starting point of the selected text, if target type is "Image", set as -1
        private Integer posEnd; --> ending point of the selected text, if  target type is "Image", set as -1
        
       - If Target type is "Images":
        private Integer imgX; --> you can use it as you wish, this is the compatible structure with w3c. if target type is "Image", -1 by default
        private Integer imgY; --> you can use it as you wish, this is the compatible structure with w3c. if target type is "Image", -1 by default
        private Integer imgW; --> you can use it as you wish, this is the compatible structure with w3c. if target type is "Image", -1  by default
        private Integer imgH; --> you can use it as you wish, this is the compatible structure with w3c. if target type is "Image", -1 by default

NOTES: 
    * we can fill all the fields, you can follow the structure as passing -1 to  unused fields. 
    * we can not update the type of an existing annotation, as a design choice

## Implemented features/endpoints:

/annotation
 
 - [x] Create/Delete/Update annotation (POST /create, /update)
      - returns status message

 - [x] Delete annotation (DELETE /delete)
      - returns status message
 
 - [x] Get annotation by annotation Id (GET /{annotation id})
     - returns Map<String, Object>

 - [x] Get all annotations (GET /all)
     - returns List<Map<String, Object>>

 - [x] Get all annotations of an article (GET /all/{articleId})
     - returns List<Map<String, Object>>

### CODE REVIEW

File: backend/TraderX/src/main/java/cmpe451/group6/rest/annotation/controller/AnnotationController.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+@Api(tags = "Annotation")
+public class AnnotationController {
+
+    @Autowired
+    AnnotationService annotationService;
+
+    @Autowired
+    private Util util;
+
+    @GetMapping("{id}")
```
Comment (enozcan) [inline backend/TraderX/src/main/java/cmpe451/group6/rest/annotation/controller/AnnotationController.java]: Is a slash needed here? I mean, is the endpoint `annotation/1` or  `annotation1` for this form?

Comment (barisegesevgili) [reply backend/TraderX/src/main/java/cmpe451/group6/rest/annotation/controller/AnnotationController.java]: `annotation/1` in this form. Most probably system auto handles this issue. I checked it out as `annotation/1` and no problems

Comment (barisegesevgili) [reply backend/TraderX/src/main/java/cmpe451/group6/rest/annotation/controller/AnnotationController.java]: `annotation/1` in this form. Most probably system auto handles this issue. I checked it out as `annotation/1` and no problems

File: backend/TraderX/src/main/java/cmpe451/group6/rest/annotation/service/AnnotationService.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+            annotation.setPosStart(-1);
+            annotation.setPosEnd(-1);
+            annotation.setImgX(annotationDTO.getImgX());
+            annotation.setImgY(annotationDTO.getImgY());
+            annotation.setImgW(annotationDTO.getImgW());
+            annotation.setImgH(annotationDTO.getImgH());
+
+        } else {
+            throw new CustomException("Target type is incorrect", HttpStatus.PRECONDITION_FAILED);
```
[the remark this answers is not in the document]
Comment (barisegesevgili) [reply backend/TraderX/src/main/java/cmpe451/group6/rest/annotation/service/AnnotationService.java]: I have contacted with them about the necessary formats, however I better change the message!

**Pull request merged**